jd4/api.py

```diff
@@ -57,10 +57,28 @@ def full_url(self, *parts):
     async def judge_consume(self, handler_type):
         async with self.ws_connect(self.full_url('judge/consume-conn/websocket')) as ws:
             logger.info('Connected')
-            async for msg in ws:
-                request = json.loads(msg.data)
-                await handler_type(self, request, ws).handle()
+            queue = Queue()
+            async def worker():
+                try:
+                    while True:
+                        request = await queue.get()
+                        await handler_type(self, request, ws).handle()
+                except CancelledError:
+                    raise
+                except Exception as e:
+                    logger.exception(e)
+                    ws.close()
+            worker_task = get_event_loop().create_task(worker())
+            try:
+                while True:
+                    await queue.put(await ws.receive_json())
+            except TypeError:
+                pass
             logger.warning('Connection lost with code %d', ws.close_code)
+            if worker_task.done():
```

Comment:
We need to wait for `worker_task` after calling `cancel()`. I would just

```
worker_task.cancel()
await worker_task
```

Reply:
but awaiting after cancelling will cause `CancelledError`?

Reply:
then we need to eat the error:

```
worker_task.cancel()
try:
    await worker_task
except CancelledError:
    pass
```

Reply:
gg, ugly